### PR TITLE
feat(events): add audit trail feature

### PR DESCRIPTION
**Pull request by sahil143**
<strong>Note: If there are pending changes to the PR, prefix the PR title with "WIP" and add the label "DO NOT MERGE"</strong>

### Mandatory
 - [ ] What does this PR do? 
 Implement Audit-Trail feature

- [ ] What issue/task does this PR references?
https://openshift.io/openshiftio/openshiftio/plan/detail/1915

- [ ] Are the tests Included?
Added tests
___
### Optional
- [ ] Is the documentation Included?

- [ ] Are the Release Notes included?

- [ ] @mention(s) to expected reviewer(s) included

**Note:**
Document for Audit-Trail messages: https://docs.google.com/spreadsheets/d/1fi2ZbAECorWkM3Krug7HhyXsUdF4wa76RWTPmPMhDkU/edit?ts=5b080e62#gid=423213162

### CODE REVIEW

File: src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.module.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import { GlobalSettings } from "../../shared/globals";
+import { HttpService } from './../../services/http-service';
+import { HttpModule, Http, XHRBackend, RequestOptions } from '@angular/http';
+import { AuthenticationService } from 'ngx-login-client';
+import { MockHttp } from '../../mock/mock-http';
+import { EventQuery } from "../../models/event.model";
+
+let providers = [];
+
+if (process.env.ENV == 'inmemory') {
```
Comment (jarifibrahim) [inline src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.module.ts]: Why `inmemory` mode?
We no longer use `inmemory` mode.

File: src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        </ng-template>
+      </a>
+    </li>
+  </ul>
+  <div [collapse]="isCollapsedTimeline">
+      <work-item-event 
+        *ngFor="let event of events | async"
+        [event]="event"
+        [type]="event.type"
+        >
```
Comment (divyanshiGupta) [inline src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.html]: You can remove this extra line and add `>` in line 28

File: src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  @Input('workItem') set WorkItem(workitem: WorkItemUI) {
+    this.workitem = workitem;
+    this.store.dispatch(new EventActions.Get(this.workitem.eventLink));
+  }
+
+  private workitem: WorkItemUI = null;
+
+  private events = this.eventsQuery.getEventsWithModifier();
+
+    constructor( private store: Store < AppState >,
```
Comment (divyanshiGupta) [inline src/app/components_ngrx/work-item-event-wrapper/work-item-event-wrapper.component.ts]: Indent the constructor properly and remove extra spaces in the parameters of the constructor, it should be
`private store: Store<AppState>`

File: src/app/components_ngrx/work-item-event/work-item-event.component.html [hunk 4]
```diff
@@ -0,0 +1,65 @@
+<div class="f8-comment__wrap">
+    <div class="user-avatar margin-right-10 pull-left">
+      <img class="user-assign-avatar"
+        placement="right"
+        container="body"
+        *ngIf="(event.modifier | async).avatar"
+        [src]="(event.modifier | async).avatar | almAvatarSize:30" />
```
Comment (divyanshiGupta) [inline src/app/components_ngrx/work-item-event/work-item-event.component.html]: remove extra space before closing the tag

File: src/app/components_ngrx/work-item-event/work-item-event.component.less [hunk 5]
```diff
@@ -0,0 +1,20 @@
+@import (reference) '../../../assets/stylesheets/_base';
+@import './../work-item-comment/work-item-comment.component.less';
+
+
```
Comment (divyanshiGupta) [inline src/app/components_ngrx/work-item-event/work-item-event.component.less]: remove extra line

File: src/app/components_ngrx/work-item-event/work-item-event.component.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    } else {
+      this.intermediateText = "added the label";
+      this.toText = " and removed the label";
+    }
+    this.textType = "label";
+  }
+
+
+
+}
```
Comment (divyanshiGupta) [inline src/app/components_ngrx/work-item-event/work-item-event.component.ts]: remove extra lines

**Comment by maxandersen** [comment]
I tried to spot it in code but couldn't - is this using the existing audit logs backend already does for all table updates implemented by @xcoulon a while back or is this introducing something new ? 

And how does https://docs.google.com/spreadsheets/d/1fi2ZbAECorWkM3Krug7HhyXsUdF4wa76RWTPmPMhDkU/edit?ts=5b080e62#gid=423213162 relate ?

**Comment by sahil143** [comment]
>I tried to spot it in code but couldn't - is this using the existing audit logs backend already does for all table  updates implemented by @xcoulon a while back or is this introducing something new ?

@maxandersen Dont know if @xcoulon implementation is being used by backend or not. But, This ([Events](https://github.com/fabric8-services/fabric8-wit/pull/1959)) was the Pr that introduced the events api.
@baijum ^

The spreadsheet contains all the text formats used to show the audit on the particular field of the work item on the UI.

**Comment by Raunak1203** [comment]
@sahil143 if i update iteration/Area from parent to child, logs shown are not correct, 
- Changed the area from 'child area' to 'new_test_space' 
but logs shows "_Raunak  moved this work item from area 'new_test_space' to area 'child area_" 

peek 2018-05-28 22-19

**Comment by sahil143** [comment]
> Shouldn't the timeline changes be collapsed bydefault(right angle) ?

we are going to merge comments and audit trail together, So it will be open by default, I have added the empty text for this one when there is no event.

> I can't see logs for Added/removed links from a workitem.

We are not capturing logs for links 

> In audit logs, Labels are clickable, If i click a label, results are filtered based on label condition.

fixed this

> It make sense if we add a label but if we remove it, it doesn't make sense.

can you explain this? I dont understand

> Changed the area from 'child area' to 'new_test_space'
but logs shows "Raunak moved this work item from area 'new_test_space' to area 'child area"

I know I already fixed this. Text should be "Raunak moved this work item to area 'new_test_space' from area 'child area"

> And also in the detail page clicking label is appending the URL but no change in the detail page as expected.

This should not happen anymore

@Raunak1203 ^^
And Please review again

**Pull request merged**